File: src/augmenters/linear_bottleneck.py
from typing_extensions import override

from src.shockwave_drawer import ShockwaveDrawer

from ..drawer_utils import CapacityEvent, UserInterface, dtPoint, float_isclose
from .base_augmenter import CapacityBottleneck


class LineBottleneck(CapacityBottleneck):
    def __init__(self, start: dtPoint, end: dtPoint, bottleneck_capacity: float):
        """Traffic light constructor.

        Args:
            pos (float): the position at which the traffic light is located (meters)
            cycles (tuple[float, ...]): the times of each traffic light cycle
            blocking_states (tuple[bool, ...]): whether or not each cycle is blocking or not
            init_state (int, optional): initial state of the traffic light. Defaults to 0.

        Raises:
            ValueError: the state must be a valid index of cycle
            ValueError: each cycle of the traffic light must be defined to be blocking or not
        """
        super().__init__(bottleneck_capacity)

        self.start = start  # where the traffic light is
        self.end = end

        if bottleneck_capacity < 0:
            raise ValueError("Capacity must be positive")

        self.bottleneck_capacity = bottleneck_capacity

    @override
    def init(self, drawer: ShockwaveDrawer):
        if float_isclose(self.start.time, self.end.time):
            return

        if self.start.time >= 0:
            cur = UserInterface(
                self.start,
                self.start.get_slope(self.end),
                self,
                self.start,
                self.end,
            )
            drawer._add_interface(cur)

            start_event = CapacityEvent(
                self.start, cur, posterior_capacity=self.bottleneck_capacity
            )
            drawer.events.add(start_event)

            end_event = CapacityEvent(self.end, cur, prior_capacity=self.bottleneck_capacity)
            drawer.events.add(end_event)


class HorizontalBottleneck(LineBottleneck):
    def __init__(self, pos: float, time_start: float, time_end: float, bottleneck_capacity: float):
        """Traffic light constructor.

        Args:
            pos (float): the position at which the traffic light is located (meters)
            cycles (tuple[float, ...]): the times of each traffic light cycle
            blocking_states (tuple[bool, ...]): whether or not each cycle is blocking or not
            init_state (int, optional): initial state of the traffic light. Defaults to 0.

        Raises:
            ValueError: the state must be a valid index of cycle
            ValueError: each cycle of the traffic light must be defined to be blocking or not
        """
        super().__init__(dtPoint(time_start, pos), dtPoint(time_end, pos), bottleneck_capacity)

File: src/augmenters/linear_bottleneck.ts
import { CapacityEvent, UserInterface, dtPoint, floatIsClose } from "../drawer_utils";
import { ShockwaveDrawer } from "../shockwave_drawer";

import { CapacityBottleneck } from "./base_augmenter";


export class LineBottleneck extends CapacityBottleneck {
  private start: dtPoint;
  private end: dtPoint;

  constructor(start: dtPoint, end: dtPoint, bottleneckCapacity: number) {
    super(bottleneckCapacity);

    this.start = start;
    this.end = end;
  }

  public init(drawer: ShockwaveDrawer): void {
    if (floatIsClose(this.start.time, this.end.time)) {
      return;
    }

    if (this.start.time >= 0) {
      const cur: UserInterface = new UserInterface(
        this.start, this.start.getSlope(this.end), this, this.start, this.end
      );

      drawer.addUserInterface(cur);

      const startEvent = new CapacityEvent(
        this.start, cur, undefined, this.bottleneck
      );
      drawer.addCapacityEvent(startEvent);

      const endEvent = new CapacityEvent(
        this.end, cur, this.bottleneck
      );
      drawer.addCapacityEvent(endEvent);
    }
  }
}

export class HorizontalBottleneck extends LineBottleneck {
  constructor(pos: number, timeStart: number, timeEnd: number, bottleneckCapacity: number) {
    super(new dtPoint(timeStart, pos), new dtPoint(timeEnd, pos), bottleneckCapacity);
  }
}